ibis/backends/dask/executor.py

```diff
@@ -0,0 +1,414 @@
+                lhs = df[left.name][mask].compute()
+                rhs = df[right.name][mask].compute()
+                return lhs.cov(rhs, ddof=ddof)
+
+        return agg
+
+    @classmethod
+    def visit(
+        cls, op: ops.ReductionVectorizedUDF, func, func_args, input_type, return_type
+    ):
+        def agg(df):
+            # if df is a dask dataframe then we collect it to a pandas dataframe
+            # because the user-defined function expects a pandas dataframe
+            if isinstance(df, dd.DataFrame):
+                df = df.compute()
+            args = [df[col.name] for col in func_args]
+            return func(*args)
+
+        return agg
+
+    @classmethod
+    def visit(
+        cls, op: ops.AnalyticVectorizedUDF, func, func_args, input_type, return_type
+    ):
+        def agg(df, order_keys):
+            # if df is a dask dataframe then we collect it to a pandas dataframe
+            # because the user-defined function expects a pandas dataframe
+            if isinstance(df, dd.DataFrame):
+                df = df.compute()
+            args = [df[col.name] for col in func_args]
+            res = func(*args)
+            if isinstance(res, pd.DataFrame):
+                # it is important otherwise it is going to fill up the memory
+                res = res.apply(lambda row: row.to_dict(), axis=1)
+            return res
+
+        return agg
+
+    ############################ Window functions #############################
+
+    @classmethod
+    def visit(cls, op: ops.WindowFrame, table, start, end, **kwargs):
+        table = table.compute()
+        if isinstance(start, dd.Series):
+            start = start.compute()
+        if isinstance(end, dd.Series):
+            end = end.compute()
+        return super().visit(op, table=table, start=start, end=end, **kwargs)
+
+    @classmethod
+    def visit(cls, op: ops.WindowFunction, func, frame):
+        result = super().visit(op, func=func, frame=frame)
+        return cls.asseries(result)
+
+    ############################ Relational ###################################
+
+    @classmethod
+    def visit(cls, op: ops.DatabaseTable, name, schema, source, namespace):
+        try:
+            return source.dictionary[name]
+        except KeyError:
+            raise UnboundExpressionError(
+                f"{name} is not a table in the {source.name!r} backend, you "
+                "probably tried to execute an expression without a data source"
+            )
+
+    @classmethod
+    def visit(cls, op: ops.InMemoryTable, name, schema, data):
+        df = data.to_frame().reset_index(drop=True)
+        return dd.from_pandas(df, npartitions=1)
+
+    @classmethod
+    def visit(cls, op: ops.DummyTable, values):
+        df, _ = cls.asframe(values)
+        return df
+
+    @classmethod
+    def visit(cls, op: PandasLimit, parent, n, offset):
+        n = n.compute().iat[0, 0]
+        offset = offset.compute().iat[0, 0]
```

Comment:
You should be able to avoid the extra compute here by doing a `map_partitions` call of a function that applies the `between` operation below. This will avoid excess compute if the values of `n`/`offset` are dependent on anything in `df`.

Something like (untested):

```python
# define this at the top-level for cheaper pickling
def limit_chunk(df, col, n, offset):
    n = n.iat[0, 0]
    offset = offset.iat[0, 0]
    if n is None:
        return df[df[col] >= offset]
    else:
        return df[df[col].between(offset, offset + n - 1)]

# then in the visit function, something like:
return df.map_partitions(limit, col=name, n=n, offset=offset, align_partitions=False, meta=df._meta).drop(columns=[name])
```

Reply:
Created a follow-up https://github.com/ibis-project/ibis/issues/8186